gcc/rust/resolve/rust-default-resolver.cc

```diff
@@ -0,0 +1,805 @@
+      stmt->accept_vis (*this);
+
+    if (expr.has_tail_expr ())
+      expr.get_tail_expr ()->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Normal, expr.get_node_id (), inner_fn);
+}
+
+void
+DefaultResolver::visit (AST::Module &module)
+{
+  auto item_fn = [this, &module] () {
+    for (auto &item : module.get_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Module, module.get_node_id (), item_fn,
+		   module.get_name ());
+}
+
+void
+DefaultResolver::visit (AST::Function &function)
+{
+  auto def_fn = [this, &function] () {
+    for (auto &param : function.get_function_params ())
+      {
+	param.get_pattern ()->accept_vis (*this);
+	param.get_type ()->accept_vis (*this);
+      }
+
+    function.get_definition ()->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Function, function.get_node_id (), def_fn);
+}
+
+void
+DefaultResolver::visit (AST::Method &method)
+{
+  auto def_fn = [this, &method] () {
+    for (auto &param : method.get_function_params ())
+      {
+	param.get_pattern ()->accept_vis (*this);
+	param.get_type ()->accept_vis (*this);
+      }
+
+    method.get_definition ()->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Function, method.get_node_id (), def_fn);
+}
+
+void
+DefaultResolver::visit (AST::ForLoopExpr &expr)
+{
+  resolver.scoped (Rib::Kind::Normal, expr.get_node_id (), [this, &expr] () {
+    expr.get_pattern ()->accept_vis (*this);
+    expr.get_iterator_expr ()->accept_vis (*this);
+    expr.get_loop_block ()->accept_vis (*this);
+  });
+}
+
+void
+DefaultResolver::visit (AST::Trait &trait)
+{
+  auto inner_fn = [this, &trait] () {
+    for (auto &item : trait.get_trait_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::TraitOrImpl, trait.get_node_id (), inner_fn,
+		   trait.get_identifier () /* FIXME: Is that valid?*/);
+}
+
+void
+DefaultResolver::visit (AST::InherentImpl &impl)
+{
+  auto inner_fn = [this, &impl] () {
+    for (auto &item : impl.get_impl_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::TraitOrImpl, impl.get_node_id (), inner_fn);
+}
+
+void
+DefaultResolver::visit (AST::TraitImpl &impl)
+{
+  auto inner_fn = [this, &impl] () {
+    for (auto &item : impl.get_impl_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::TraitOrImpl, impl.get_node_id (), inner_fn);
+}
+
+void
+DefaultResolver::visit (AST::ExternBlock &block)
+{
+  auto inner_fn = [this, &block] () {
+    for (auto &item : block.get_extern_items ())
+      item->accept_vis (*this);
+  };
+
+  resolver.scoped (Rib::Kind::Normal /* FIXME: Correct? */,
+		   block.get_node_id (), inner_fn);
+}
+
+void
+DefaultResolver::visit (AST::StructStruct &type)
+{
+  // do we need to scope anything here? no, right?
+
+  // we also can't visit `StructField`s by default, so there's nothing to do -
+  // correct? or should we do something like
+
+  for (auto &field : type.get_fields ())
```

Comment:
Should we really visit those struct ?
```rust
struct Thing {
    a : usize,
}
```
`a` being resolved like any other name make sense in some way but it feels weird ?